Add Nios-II QSPI Flash driver

ramakrishnapallala opened the pull request.

andrewboie commented on drivers/flash/soc_flash_nios2_qspi.c:
```diff
@@ -0,0 +1,449 @@
+		}
+
+		/* prepare the word to be written */
+		memcpy(((void *)&word_to_write) + padding,
```
you dont need these casts to void *

ramakrishnapallala replied:
got it.

andrewboie commented on drivers/flash/soc_flash_nios2_qspi.c:
```diff
@@ -0,0 +1,449 @@
+
+	while (remaining_length > 0) {
+		/* initialize word to write to blank word */
+		word_to_write = 0xFFFFFFFF;
```
consider #defining this value

ramakrishnapallala replied:
Ok

andrewboie commented on drivers/flash/soc_flash_nios2_qspi.c:
```diff
@@ -0,0 +1,449 @@
+	if ((data == NULL) || ((offset + len) > qspi_dev->data_end) ||
+			(0 != (write_offset &
+			       (NIOS2_WRITE_BLOCK_SIZE - 1)))) {
+		printk("nios2-qspi: read failed at offset %ld\n", offset);
```
consider using Zephyr's error logging macros instead of printk() directly

ramakrishnapallala replied:
Ok.

andrewboie commented on tests/boards/altera_max10/qspi/testcase.yaml:
```diff
@@ -0,0 +1,4 @@
+tests:
+  test:
+    platform_whitelist: altera_max10
```
Will this work on any MAX10 device, are there any pre-conditions on its setup?

ramakrishnapallala replied:
This should work on any max10 device provided it's softcpu has support for QSPI soft IP

tyrantlizard77 reviewed (commented):
Looks pretty good just nitpicky stuff

ramakrishnapallala commented:
@andrewboie I have fixed your comment and updated the PR. Pls have a look it.

ramakrishnapallala commented:
@andrewboie I have updated the patches now, can you have a look at it.

nashif commented:
@ramakrishnapallala  please fix conflict

Commit:
ext: hal: altera: Add QSPI Controller HAL driver

Add Altera Nios-II QSPI Controller HAL driver to Zephyr.
The sources are taken from the Altera SDK v17.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
drivers: flash: Add Nios-II QSPI flash driver

Add Altera Nios-II QSPI Flash controller driver which has
has 1024 blocks or sectors wich each sector size being 64K bytes.
This driver supports flash erase, write, read and lock operations.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
tests: boards: Add QSPI flash test app for altera_max10

Add test application to demonstrate the usage of Nios-II
QSPI Flash soft IP.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

ramakrishnapallala commented:
@nashif fixed the conflict.

@andrewboie @lpereira please have a look at the updated patches.

rgundi reviewed (commented):
Code looks good to me
